### ISSUE TITLE
logsigmoid fails on -inf

### ISSUE THREAD

**Issue by drewsil** [pytorch/pytorch#20972]
## 🐛 Bug

torch.nn.functional.logsigmoid returns nan when its input is -inf, the correct return value should be -inf. Algorithm should be adjusted to return the correct value (and correct derivative), or the special case should be tested for and replaced.

The impact of the issue is that if a -inf is present in an input tensor the output is corrupted and the gradient will fail to compute for the full tensor, even though -inf should not be problematic.

In fixing the error the gradient should also be checked and should be 1.0 (currently gradient calculation fails due to nan.)

## To Reproduce

1. Run the code snippet below and get a nan result.
```python
>>> x = torch.tensor(-float('Inf'),requires_grad=True)
>>> y = torch.nn.functional.logsigmoid(x)
>>> y.backward()
>>> print('y = '+str(y.item()))
y = nan
>>> print('dy/dx = '+str(x.grad.item()))
dy/dx = nan
```

## Expected behavior
```python
>>> x = torch.tensor(-float('Inf'),requires_grad=True)
>>> y = torch.nn.functional.logsigmoid(x)
>>> y.backward()
>>> print('y = '+str(y.item()))
y = -inf
>>> print('dy/dx = '+str(x.grad.item()))
dy/dx = 1.0
```

## Environment
 - PyTorch Version (e.g., 1.0): 1.1 (cpu only)
 - OS (e.g., Linux): Windows
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source): n/a
 - Python version:3.7.1
 - CUDA/cuDNN version: n/a
 - GPU models and configuration: n/a
 - Any other relevant information: n/a

## Additional context

Problem should occur generally as it is a flaw in the algorithm, I have not tested on a gpu, only on a cpu.

**Comment by odysseus9672** [comment]
This is still in issue in Pytorch 1.7.0.

**Comment by haotang1995** [comment]
The problem still exists in Pytorch 1.10.1.

**Comment by peterbell10** [comment]
The commit that fixes this isn't in 1.10. You need to wait until a new major release, or try the pytorch-nightly builds.

**Comment by haotang1995** [comment]
> The commit that fixes this isn't in 1.10. You need to wait until a new major release, or try the pytorch-nightly builds.

Thanks for the reply! I am now trying to avoid these issues by python wrappers, e.g., using nan_to_num. Hoping it will work or I will try the pytorch-nightly builds.
